File: src/tienda/Tienda.java
package tienda;

import productos.*;
import usuarios.Cliente;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Scanner;

public class Tienda {
    private static final String password ="123";
    public static Scanner sc = new Scanner(System.in);

    public static ArrayList<Electrodomestico> electrodomesticos = new ArrayList<>();
    public static ArrayList<Limpieza> limpiezas = new ArrayList<>();
    public static ArrayList<Belleza> bellezas = new ArrayList<>();
    public static ArrayList<Alimento> alimentos = new ArrayList<>();
    public static ArrayList<Cliente> clientes = new ArrayList<>();
    public static ArrayList<Producto> productos = new ArrayList<>();


//Productos

    //Electrodomestico
    public static void addElectrodomestico(Electrodomestico electrodomestico){
        electrodomesticos.add(electrodomestico);
        productos.add(electrodomestico);
        System.out.println("Producto de Electrodomestico añdadido correcatamente");
    }
    public static void mostrarElectrodomestico(){
        System.out.println("*** Productos de Electrodomestico ***");
        for (int i = 0; i < electrodomesticos.size(); i++) {
            System.out.println("["+i+"]"+"- Nombre: "+electrodomesticos.get(i).getNombre()+", Precio: "+electrodomesticos.get(i).getPrecio()+", Fechade importacion: "+ LocalDate.now()+", Numero Serie: "+electrodomesticos.get(i).getNumeroSerie()+", StockDisponible: "+electrodomesticos.get(i).getStock()+", Marca: "+electrodomesticos.get(i).getMarca());
        }
        System.out.println();
    }

    //Limpieza
    public static void addLimpieza(Limpieza limpieza){
        limpiezas.add(limpieza);
        productos.add(limpieza);
        System.out.println("Producto de Limpieza añdadido correcatamente");

    }

    public static void mostrarLimpieza() {
        System.out.println("*** Productos de Limpieza ***");
        for (int i = 0; i < limpiezas.size(); i++) {
            System.out.println("["+i+"]" + "- Nombre: " + limpiezas.get(i).getNombre() + ", Precio: " + limpiezas.get(i).getPrecio() + ", Fecha de importación: " + LocalDate.now() + ", Numero Serie: " + limpiezas.get(i).getNumeroSerie() + ", Stock Disponible: " + limpiezas.get(i).getStock()+", Aroma: "+limpiezas.get(i).getAroma());
        }
        System.out.println();
    }
    //Belleza
    public static void addBelleza(Belleza belleza){
        bellezas.add(belleza);
        productos.add(belleza);
        System.out.println("Producto de Belleza añdadido correcatamente");
    }

    public static void mostrarBelleza() {
        System.out.println("*** Productos de Belleza ***");
        for (int i = 0; i < bellezas.size(); i++) {
            System.out.println("["+i+"]"+ "- Nombre: " + bellezas.get(i).getNombre() + ", Precio: " + bellezas.get(i).getPrecio() + ", Fecha de importación: " + LocalDate.now() + ", Numero Serie: " + bellezas.get(i).getNumeroSerie() + ", Stock Disponible: " + bellezas.get(i).getStock()+", Tono de Piel: "+bellezas.get(i).getTonoPiel());
        }
        System.out.println();
    }
    //Alimento
    public static void addAlimeto(Alimento alimento){
        alimentos.add(alimento);
        productos.add(alimento);
        System.out.println("Producto de Alimento añdadido correcatamente");
    }
    public static void mostrarAlimento() {
        System.out.println("*** Productos de Alimentos ***");
        for (int i = 0; i < alimentos.size(); i++) {
            System.out.println("["+i+"]"+ "- Nombre: " + alimentos.get(i).getNombre() + ", Precio: " + alimentos.get(i).getPrecio() + ", Fecha de importación: " + LocalDate.now() + ", Numero Serie: " + alimentos.get(i).getNumeroSerie() + ", Stock Disponible: " + alimentos.get(i).getStock()+", Fecha de Caducidad: "+alimentos.get(i).getFechaCaducidad());
        }
        System.out.println();
    }


    public static void mostrarTodosLosProductosPorCategoria() {
        mostrarAlimento();
        mostrarBelleza();
        mostrarElectrodomestico();
        mostrarLimpieza();
    }
    public static void mostrarTodosLosProductos() {
        for (Producto producto : productos) {
            if (!producto.isComprado() || producto.getStock()>0) {
                System.out.println( "["+productos.indexOf(producto)+"]" + producto.getNombre() + "\n" + "Precio: " + producto.getPrecio() + ", Fecha de Importacion: " + producto.getFechaImportacion() + "\n" + "Cantidad en stock: " + producto.getStock() + ", Esta comprado? " + producto.isComprado() + ", Numero de serie: " + producto.getNumeroSerie() + "\n");
            }
        }
    }
    public static void removerProducto(int indexRemover){
       Producto producto=productos.get(indexRemover);
       if (!producto.isComprado()){
           productos.remove(indexRemover);
           System.out.println("Producto Removido Exitosamente");
           if (alimentos.contains(producto)){
               alimentos.remove(producto);
           } else if (bellezas.contains(producto)) {
               bellezas.remove(producto);
           } else if (electrodomesticos.contains(producto)) {
               electrodomesticos.remove(producto);
           } else if (limpiezas.contains(producto)) {
               limpiezas.contains(producto);
           }
       }else if (producto.isComprado()){
           System.out.println("No se puede eliminar el producto, el producto esta asociado a compras de los clientes");
       }else{
           System.out.println("indice invalido");
       }

    }

//Cliente
    public void addCliente(Cliente cliente) {
        clientes.add(cliente);
        System.out.println("Cliente agregado correctamente. ");
    }


    public static void removerCliente(int indexRemover){
        Cliente cliente=clientes.get(indexRemover);
        if (cliente.getCompras().isEmpty()){
            clientes.remove(indexRemover);
            System.out.println("Cliente Removido Exitosamente");
        }else{
            System.out.println("No se puede eliminar el cliente, el cliente tiene compras asociadas");
        }

    }
    public static void mostrarCliente(){
        System.out.println("*** Clientes ***");
        for (int i = 0; i <clientes.size() ; i++) {
            System.out.println("["+i+"]- "+clientes.get(i).getName()+" "+clientes.get(i).getApellido());
        }
    }

    public static void numeroClientes(){
        System.out.println("El numero de clientes es: "+clientes.size());
    }
    public static void numeroProductos(){
        System.out.println("El numero de productos es: "+productos.size());
    }


    public static boolean checkPaswword(String contra){
        return contra.equals(password);
    }

    //comprar productos
    public static void comprarProducto(int indexCliente,int indexProducto ){
        Producto producto= productos.get(indexProducto);

        if  (producto.isComprado() && producto.getStock()<1){
            System.out.println("Producto no disponible");
        }
        else if (producto.getStock()>0){

            producto.setComprado();
            producto.removerStock(1);
            Cliente cliente=clientes.get(indexCliente);
            LocalDate fecha = LocalDate.now();
            cliente.anadirProducto(producto,fecha);
            System.out.println("!Producto comprado¡");
        }
        else{
            System.out.println("No hay stock disponible");
        }

    }
    public static void productosComprados(){
        for (int i = 0; i < clientes.size(); i++) {
            Cliente cliente=clientes.get(i);
            System.out.println("Compras realizadas por: " + cliente.getName()+" "+cliente.getApellido());
            cliente.mostrarCompras();
            System.out.println();
        }
    }

}

File: src/tienda/Tienda.py
from datetime import datetime


class Tienda:
    __password = "123"
    __electrodomesticos = []
    __limpiezas = []
    __bellezas = []
    __alimentos = []
    __clientes = []
    __productos = []
    __fecha_actual = datetime.now()

    # Electrodomestico
    def add_electrodomestico(self, electrodomestico):
        self.__electrodomesticos.append(electrodomestico)
        self.__productos.append(electrodomestico)
        print("Producto de Electrodomestico añdadido correcatamente")

    def mostrar_electrodomestico(self):
        print("*** Productos de Electrodomestico ***")
        for i, electrodomestico in enumerate(self.__electrodomesticos):
            print(f"[{i}] - Nombre: {electrodomestico.get_nombre()}, "
                  f"Precio: {electrodomestico.get_precio()}, "
                  f"Fecha de importacion: {self.__fecha_actual}, "
                  f"Número de Serie: {electrodomestico.get_numero_serie()}, "
                  f"Stock Disponible: {electrodomestico.get_stock()}, "
                  f"Marca: {electrodomestico.get_marca()}")
            print()

    # Limpieza
    def add_limpieza(self, limpieza):
        self.__limpiezas.append(limpieza)
        self.__productos.append(limpieza)
        print("Producto de Limpieza añdadido correcatamente")

    def mostrar_limpiezas(self):
        print("*** Productos de Limpieza ***")
        for i, limpieza in enumerate(self.__limpiezas):
            print(f"[{i}] - Nombre: {limpieza.get_nombre()}, "
                  f"Precio: {limpieza.get_precio()}, "
                  f"Fecha de importación: {self.__fecha_actual}, "
                  f"Número de Serie: {limpieza.get_numero_serie()}, "
                  f"Stock Disponible: {limpieza.get_stock()}, "
                  f"Aroma: {limpieza.get_aroma()}")
            print()

    # Belleza

    def add_belleza(self, belleza):
        self.__bellezas.append(belleza)
        self.__productos.append(belleza)
        print("Producto de Belleza añdadido correcatamente")

    def mostrar_belleza(self):
        print("*** Productos de Belleza ***")
        for i, belleza in enumerate(self.__bellezas):
            print(f"[{i}] - Nombre: {belleza.get_nombre()}, "
                  f"Precio: {belleza.get_precio()}, "
                  f"Fecha de importación: {self.__fecha_actual}, "
                  f"Número de Serie: {belleza.get_numero_serie()}, "
                  f"Stock Disponible: {belleza.get_stock()}, "
                  f"Tono de Piel: {belleza.get_tono_piel()}")
            print()

    # Alimento
    def add_alimento(self, alimento):
        self.__alimentos.append(alimento)
        self.__productos.append(alimento)
        print("Producto de Alimento añdadido correcatamente")

    def mostrar_alimento(self):
        print("*** Productos de Alimentos ***")
        for i, alimento in enumerate(self.__alimentos):
            print(f"[{i}] - Nombre: {alimento.get_nombre()}, "
                  f"Precio: {alimento.get_precio()}, "
                  f"Fecha de importación: {self.__fecha_actual}, "
                  f"Número de Serie: {alimento.get_numero_serie()}, "
                  f"Stock Disponible: {alimento.get_stock()}, "
                  f"Fecha de Caducidad: {alimento.get_fecha_caducidad()}")
            print()

    def mostrar_todos_los_productos_por_categoria(self):
        self.mostrar_belleza()
        self.mostrar_alimento()
        self.mostrar_electrodomestico()
        self.mostrar_limpiezas()

    def mostrar_todos_los_productos(self):
        for i, producto in enumerate(self.__productos):
            if not producto.get_comprado() or producto.get_stock() > 0:
                print(f"[{i}] {producto.get_nombre()}\n"
                      f"Precio: {producto.get_precio()}, Fecha de Importación: {producto.get_fecha_importacion()}\n"
                      f"Cantidad en stock: {producto.get_stock()}, ¿Está comprado? {producto.is_comprado()}"
                      f", Número de serie: {producto.get_numero_serie()}\n")

    # Remover Producto

    def remover_producto(self, index_remover):
        producto = self.__productos[index_remover]
        if not producto.get_comprado():
            self.__productos.pop(index_remover)
            print("Producto Removido Exitosamente")
            if producto in self.__alimentos:
                self.__alimentos.remove(producto)
            elif producto in self.__bellezas:
                self.__bellezas.remove(producto)
            elif producto in self.__electrodomesticos:
                self.__electrodomesticos.remove(producto)
            elif producto in self.__limpiezas:
                self.__limpiezas.remove(producto)
        elif producto.get_comprado():
            print("No se puede eliminar el producto, el producto esta asociado a compras de los clientes")
        else:
            print("indice invalido")

    # Cliente

    def add_cliente(self, cliente):
        self.__clientes.append(cliente)
        print("Cliente agregado correctamente. ")

    def remove_cliente(self, index_remover):
        cliente = self.__clientes[index_remover]
        if not cliente.get_comprado():
            self.__clientes.pop(index_remover)
            print("Cliente Removido Exitosamente")
        else:
            print("No se puede eliminar el cliente, el cliente tiene compras asociadas")

    def get_cliente(self, index):
        if 0 <= index < len(self.__clientes):
            return self.__clientes[index]
        else:
            return None

    def mostrar_clientes(self):
        print("*** Clientes ***")
        for i, cliente in enumerate(self.__clientes):
            print(f"[{i}] - {cliente.get_name()} {cliente.get_apellido()}")

    def numero_clientes(self):
        print("El numero de clientes es: ", len(self.__clientes))

    def numero_productos(self):
        print("El numero de productos es: ", len(self.__productos))

    def check_paswword(self, password):
        if password == self.__password:
            return True
        else:
            return False

    # comprar productos
    def comprar_productos(self, index_cliente, index_producto):
        producto = self.__productos[index_producto]
        if producto.get_comprado() and producto.get_stock() < 1:
            print("Producto no disponible")
        elif producto.get_stock() > 0:
            producto.set_stock()
            producto.remover_stock(self, 1)
            cliente = self.__clientes[index_cliente]
            cliente.add_producto(producto, self.__fecha_actual)
            print("!Producto comprado¡")
        else:
            print("No hay stock disponible")

    def productos_comprados(self):
        for cliente in self.__clientes:
            print("Compras realizadas por:", cliente.get_name(), cliente.get_apellido())
            cliente.mostrar_compras()
            print()
